File: code/ch16/16.4.1.permutations.java
class Solution {
    List<List<Integer>> ans = new ArrayList<>();
    Set<Integer> visited = new HashSet<>();
    public List<List<Integer>> permute(int[] nums) {
        dfs(new ArrayList<>(), nums);
        return ans;
    }
    public void dfs(List<Integer> path, int[] nums) {
        if (path.size() == nums.length) {
            ans.add(new ArrayList<>(path));
            return;
        }
        for (int i = 0; i < nums.length; i++) {
            if (!visited.contains(i)) {
                visited.add(i);
                path.add(nums[i]);
                dfs(path, nums);
                path.remove(path.size() - 1);
                visited.remove(i);
            }
        }
    }
}

File: code/ch16/16.4.1.permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []
        n = len(nums)
        visited = set()

        def dfs(path: List[int]):
            if len(path) == n:
                ans.append(path[:])
                return
            for i in range(n):
                if i not in visited:
                    visited.add(i)
                    path.append(nums[i])
                    dfs(path)
                    path.pop()
                    visited.remove(i)

        dfs([])
        return ans
